### PR TITLE
use the provided system rocksdb prebuilt on freebsd

### PR DESCRIPTION
**Pull request by girlbossceo**
Without this, it causes an unnecessary build of RocksDB that seems to never work.

<details>
<summary>Sample of build errors</summary>

```
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.cc:6:
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.h:11:
  cargo:warning=rocksdb/port/jemalloc_helper.h:62:61: error: use of undeclared identifier 'nothrow'; did you mean 'std::nothrow'?
  cargo:warning=   62 | extern "C" JEMALLOC_ALLOCATOR JEMALLOC_RESTRICT_RETURN void JEMALLOC_NOTHROW *
  cargo:warning=      |                                                             ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:40: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                                        ^
  cargo:warning=/usr/include/c++/v1/new:126:50: note: 'std::nothrow' declared here
  cargo:warning=  126 | extern _LIBCPP_EXPORTED_FROM_ABI const nothrow_t nothrow;
  cargo:warning=      |                                                  ^
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.cc:6:
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.h:11:
  cargo:warning=rocksdb/port/jemalloc_helper.h:62:61: error: variable has incomplete type 'void'
  cargo:warning=   62 | extern "C" JEMALLOC_ALLOCATOR JEMALLOC_RESTRICT_RETURN void JEMALLOC_NOTHROW *
  cargo:warning=      |                                                             ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:26: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                          ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:62:77: error: expected ';' after top level declarator
  cargo:warning=   62 | extern "C" JEMALLOC_ALLOCATOR JEMALLOC_RESTRICT_RETURN void JEMALLOC_NOTHROW *
  cargo:warning=      |                                                                             ^
  cargo:warning=      |                                                                             ;
  cargo:warning=rocksdb/port/jemalloc_helper.h:65:61: error: use of undeclared identifier 'nothrow'; did you mean 'std::nothrow'?
  cargo:warning=   65 | extern "C" JEMALLOC_ALLOCATOR JEMALLOC_RESTRICT_RETURN void JEMALLOC_NOTHROW *
  cargo:warning=      |                                                             ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:40: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                                        ^
  cargo:warning=/usr/include/c++/v1/new:126:50: note: 'std::nothrow' declared here
  cargo:warning=  126 | extern _LIBCPP_EXPORTED_FROM_ABI const nothrow_t nothrow;
  cargo:warning=      |                                                  ^
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.cc:6:
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.h:11:
  cargo:warning=rocksdb/port/jemalloc_helper.h:65:61: error: variable has incomplete type 'void'
  cargo:warning=   65 | extern "C" JEMALLOC_ALLOCATOR JEMALLOC_RESTRICT_RETURN void JEMALLOC_NOTHROW *
  cargo:warning=      |                                                             ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:26: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                          ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:65:77: error: expected ';' after top level declarator
  cargo:warning=   65 | extern "C" JEMALLOC_ALLOCATOR JEMALLOC_RESTRICT_RETURN void JEMALLOC_NOTHROW *
  cargo:warning=      |                                                                             ^
  cargo:warning=      |                                                                             ;
  cargo:warning=rocksdb/port/jemalloc_helper.h:67:19: error: use of undeclared identifier 'nothrow'; did you mean 'std::nothrow'?
  cargo:warning=   67 | extern "C" size_t JEMALLOC_NOTHROW xallocx(void *, size_t, size_t, int)
  cargo:warning=      |                   ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:40: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                                        ^
  cargo:warning=/usr/include/c++/v1/new:126:50: note: 'std::nothrow' declared here
  cargo:warning=  126 | extern _LIBCPP_EXPORTED_FROM_ABI const nothrow_t nothrow;
  cargo:warning=      |                                                  ^
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.cc:6:
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.h:11:
  cargo:warning=rocksdb/port/jemalloc_helper.h:67:19: error: no viable conversion from 'const nothrow_t' to 'size_t' (aka 'unsigned long')
  cargo:warning=   67 | extern "C" size_t JEMALLOC_NOTHROW xallocx(void *, size_t, size_t, int)
  cargo:warning=      |                   ^~~~~~~~~~~~~~~~
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:26: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                          ^             ~~~~~~~
  cargo:warning=rocksdb/port/jemalloc_helper.h:67:35: error: expected ';' after top level declarator
  cargo:warning=   67 | extern "C" size_t JEMALLOC_NOTHROW xallocx(void *, size_t, size_t, int)
  cargo:warning=      |                                   ^
  cargo:warning=      |                                   ;
  cargo:warning=rocksdb/port/jemalloc_helper.h:69:19: error: use of undeclared identifier 'nothrow'; did you mean 'std::nothrow'?
  cargo:warning=   69 | extern "C" size_t JEMALLOC_NOTHROW sallocx(const void *, int)
  cargo:warning=      |                   ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:40: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                                        ^
  cargo:warning=/usr/include/c++/v1/new:126:50: note: 'std::nothrow' declared here
  cargo:warning=  126 | extern _LIBCPP_EXPORTED_FROM_ABI const nothrow_t nothrow;
  cargo:warning=      |                                                  ^
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.cc:6:
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.h:11:
  cargo:warning=rocksdb/port/jemalloc_helper.h:69:19: error: redefinition of 'JEMALLOC_ATTR'
  cargo:warning=   69 | extern "C" size_t JEMALLOC_NOTHROW sallocx(const void *, int)
  cargo:warning=      |                   ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:26: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                          ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:67:19: note: previous definition is here
  cargo:warning=   67 | extern "C" size_t JEMALLOC_NOTHROW xallocx(void *, size_t, size_t, int)
  cargo:warning=      |                   ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:26: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                          ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:69:35: error: expected ';' after top level declarator
  cargo:warning=   69 | extern "C" size_t JEMALLOC_NOTHROW sallocx(const void *, int)
  cargo:warning=      |                                   ^
  cargo:warning=      |                                   ;
  cargo:warning=rocksdb/port/jemalloc_helper.h:71:17: error: use of undeclared identifier 'nothrow'; did you mean 'std::nothrow'?
  cargo:warning=   71 | extern "C" void JEMALLOC_NOTHROW dallocx(void *, int) __attribute__((__weak__));
  cargo:warning=      |                 ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:40: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                                        ^
  cargo:warning=/usr/include/c++/v1/new:126:50: note: 'std::nothrow' declared here
  cargo:warning=  126 | extern _LIBCPP_EXPORTED_FROM_ABI const nothrow_t nothrow;
  cargo:warning=      |                                                  ^
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.cc:6:
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.h:11:
  cargo:warning=rocksdb/port/jemalloc_helper.h:71:17: error: variable has incomplete type 'void'
  cargo:warning=   71 | extern "C" void JEMALLOC_NOTHROW dallocx(void *, int) __attribute__((__weak__));
  cargo:warning=      |                 ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:26: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                          ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:71:33: error: expected ';' after top level declarator
  cargo:warning=   71 | extern "C" void JEMALLOC_NOTHROW dallocx(void *, int) __attribute__((__weak__));
  cargo:warning=      |                                 ^
  cargo:warning=      |                                 ;
  cargo:warning=rocksdb/port/jemalloc_helper.h:72:17: error: use of undeclared identifier 'nothrow'; did you mean 'std::nothrow'?
  cargo:warning=   72 | extern "C" void JEMALLOC_NOTHROW sdallocx(void *, size_t, int)
  cargo:warning=      |                 ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:40: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                                        ^
  cargo:warning=/usr/include/c++/v1/new:126:50: note: 'std::nothrow' declared here
  cargo:warning=  126 | extern _LIBCPP_EXPORTED_FROM_ABI const nothrow_t nothrow;
  cargo:warning=      |                                                  ^
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.cc:6:
  cargo:warning=In file included from rocksdb/memory/jemalloc_nodump_allocator.h:11:
  cargo:warning=rocksdb/port/jemalloc_helper.h:72:17: error: variable has incomplete type 'void'
  cargo:warning=   72 | extern "C" void JEMALLOC_NOTHROW sdallocx(void *, size_t, int)
  cargo:warning=      |                 ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:26: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                          ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:72:33: error: expected ';' after top level declarator
  cargo:warning=   72 | extern "C" void JEMALLOC_NOTHROW sdallocx(void *, size_t, int)
  cargo:warning=      |                                 ^
  cargo:warning=      |                                 ;
  cargo:warning=rocksdb/port/jemalloc_helper.h:74:19: error: use of undeclared identifier 'nothrow'; did you mean 'std::nothrow'?
  cargo:warning=   74 | extern "C" size_t JEMALLOC_NOTHROW nallocx(size_t, int) JEMALLOC_ATTR(pure)
  cargo:warning=      |                   ^
  cargo:warning=rocksdb/port/jemalloc_helper.h:50:40: note: expanded from macro 'JEMALLOC_NOTHROW'
  cargo:warning=   50 | #define JEMALLOC_NOTHROW JEMALLOC_ATTR(nothrow)
  cargo:warning=      |                                        ^
  cargo:warning=/usr/include/c++/v1/new:126:50: note: 'std::nothrow' declared here
  cargo:warning=  126 | extern _LIBCPP_EXPORTED_FROM_ABI const nothrow_t nothrow;
  cargo:warning=      |                                                  ^
  cargo:warning=fatal error: too many errors emitted, stopping now [-ferror-limit=]
  cargo:warning=20 errors generated.
```

</details>

FreeBSD user @Aranjedeath confirms that this patch has fixed RocksDB usage in my Rust crate.

`ldd` output of binary showing it use the system RocksDB prebuilt:
```
    Finished `release` profile [optimized] target(s) in 5m 08s
➜  aranjeconduwuit git:(main) ✗ ldd ./target/release/conduit
./target/release/conduit:
        librocksdb.so.9.4 => /usr/local/lib/librocksdb.so.9.4 (0x31a997200000)     <==============
        libthr.so.3 => /lib/libthr.so.3 (0x31a992335000)
        libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x31a995986000)
        libc.so.7 => /lib/libc.so.7 (0x31a996506000)
        libm.so.5 => /lib/libm.so.5 (0x31a99873e000)
        libsnappy.so.1 => /usr/local/lib/libsnappy.so.1 (0x31a999727000)
        libgflags.so.2.2 => /usr/local/lib/libgflags.so.2.2 (0x31a99a0d8000)
        libz.so.6 => /lib/libz.so.6 (0x31a99aeb7000)
        libbz2.so.4 => /usr/lib/libbz2.so.4 (0x31a99b328000)
        liblz4.so.1 => /usr/local/lib/liblz4.so.1 (0x31a99ce19000)
        libzstd.so.1 => /usr/local/lib/libzstd.so.1 (0x31a99b781000)
        libexecinfo.so.1 => /usr/lib/libexecinfo.so.1 (0x31a99bef6000)
        libc++.so.1 => /lib/libc++.so.1 (0x31a99d7b7000)
        libcxxrt.so.1 => /lib/libcxxrt.so.1 (0x31a99ec1c000)
        libelf.so.2 => /lib/libelf.so.2 (0x31a99dd61000)
        [vdso] (0x31a991368000)
```

cc @zaidoon1

### CODE REVIEW

**Comment by zaidoon1** [comment]
Disclaimer: I haven't tried this myself

[In https://github.com/rust-rocksdb/rust-rocksdb/pull/166](https://github.com/rust-rocksdb/rust-rocksdb/pull/166) introduced the ability to link to system rocksdb by specifying env vars, does this not work for you? Maybe we just need to document this better if it does work and you didn't know it existed? 


Side note, there has been a request to make it easier to link to system rocksdb, see https://github.com/rust-rocksdb/rust-rocksdb/issues/310 and some work that has been started (but unfortunately abandoned): https://github.com/rust-rocksdb/rust-rocksdb/pull/354


If the current env vars way doesn't work for you, maybe you can pull in some of the changes from the incomplete pr above, specfiically:
```
Feature-based selection of linking mode (static or dynamic)
Feature-based selection of vendoring (defaults to on)
```

**Comment by girlbossceo** [comment]
> https://github.com/rust-rocksdb/rust-rocksdb/pull/166 introduced the ability to link to system rocksdb by specifying env vars, does this not work for you? Maybe we just need to document this better if it does work and you didn't know it existed?

It does work, however I don't feel like this is really something that the user is expected to be doing when they use *any* Rust project with rust-rocksdb. That's the problem this PR is trying to fix.

Their situation was doing a normal `cargo build --release` on my Rust crate would never work, and it uses RocksDB via rust-rocksdb. They identified the environment variable today after dealing with all sorts of troubleshooting for days, and that totally works, but this shouldn't be a step any FreeBSD user is required to do just because a crate happens to use rust-rocksdb.

**Comment by zaidoon1** [comment]
fair point

**Comment by girlbossceo** [comment]
And for the record, I do agree that a more universal solution to using the system RocksDB is definitely warranted in the long run as ad-hoc entries in here is not sustainable, but this is definitely something trivial to fix/add in the mean time, so users don't need to locate an additional hidden step to build any Rust crate with rust-rocksdb.

File: librocksdb-sys/build.rs [hunk 1]
```diff
@@ -359,13 +359,26 @@ fn main() {
...
+        // we dont need to rebuild rocksdb
+        if target.contains("freebsd") {
+            println!("cargo:rustc-link-search=native=/usr/local/lib");
+            let mode = match env::var_os("ROCKSDB_STATIC") {
+                Some(_) => "static",
+                None => "dylib",
+            };
+            println!("cargo:rustc-link-lib={}=rocksdb", mode);
+
+            return;
```
Comment (zaidoon1) [inline librocksdb-sys/build.rs:375]: given that we return here, we won't try to build snappy https://github.com/rust-rocksdb/rust-rocksdb/pull/908/files#diff-7377ccc9f5cb3386318bd8afcd84814e7dd6d9b40efa909fb2dc218ecb779499R389 is this expected?

Comment (girlbossceo) [reply librocksdb-sys/build.rs:375]: Oops probably not. Though I did see snappy in their `ldd` output. I can still move it around and ask them to try building again, but we do need to return somewhere before the `build_rocksdb();`.

Comment (Aranjedeath) [reply librocksdb-sys/build.rs:375]: Ya, check here: https://www.freshports.org/databases/rocksdb/

On Library dependencies section, snappy is listed and rocksdb won't install without it

**Review by zaidoon1 — APPROVED** [review]

**Pull request merged**